README.rst
Used publicly available data from:

    | **Quantitative assessment of the robustness of next-generation sequencing
      of antibody variable gene repertoires from immunized mice.**
    | Greiff, V. et al.
    | *BMC Immunol. 2014. 15(1):40. doi:10.1186/s12865-014-0040-5.*

Which may be downloaded from the EBI European Nucleotide Archive under
accession ID: ERP003950. For this example, the first 25,000
sequences of sample Replicate-1-M1 (accession: ERR346600) were used, which may
downloaded using fastq-dump.

    fastq-dump --split-files -X 25000 ERR346600

Primers sequences are available in additional file 1 of the publication and included here.

M1_atleast-2.fasta was used as input to HighVQuest. 

Mouse reference sequences are included as IMGT_MOUSE_IGH[VDJ].fasta files.

ChangeO commands used:
MakeDb.py imgt -i Greiff2014.txz -s /home/ian/NextGenSeq/Greiff2014/output/M1_atleast-2.fasta --regions --scores
ParseDb.py select -d Greiff2014_db-pass.tab -f FUNCTIONAL -u T
ParseDb.py fasta -d Greiff2014_db-pass_parse-select.tab --if SEQUENCE_ID --sf SEQUENCE_IMGT --mf V_CALL DUPCOUNT
DefineClones.py bygroup -d Greiff2014_db-pass_parse-select.tab --act set --model m1n --sym min --norm len --dist 0.25
CreateGermlines.py -d ./HighVQuestOutput/Greiff2014_db-pass_parse-select_clone-pass.tab -r IMGT_MOUSE_IGH[VDJ].fasta -g dmask --cloned


output/history.txt
bash ExamplePipeline_Greiff2014_fasta.sh 
MakeDb.py imgt -i Greiff2014.txz -s /home/ian/NextGenSeq/Greiff2014/output/M1_atleast-2.fasta --regions --scores
ParseDb.py select -d Greiff2014_db-pass.tab -f FUNCTIONAL -u T
ParseDb.py fasta -d Greiff2014_db-pass_parse-select.tab --if SEQUENCE_ID --sf SEQUENCE_IMGT --mf V_CALL DUPCOUNT
DefineClones.py bygroup -d Greiff2014_db-pass_parse-select.tab --act set --model m1n --sym min --norm len --dist 0.25
CreateGermlines.py -d ./HighVQuestOutput/Greiff2014_db-pass_parse-select_clone-pass.tab -r IMGT_MOUSE_IGH[VDJ].fasta -g dmask --cloned



Pipeline_Greiff2014.sh
#!/usr/bin/env bash
# Example pRESTO pipeline for Stranded MiSeq 2x250 data.
# Data from Greiff et al, 2014, BMC Immunol.
# 
# Author:  Jason Anthony Vander Heiden
# Date:    2016.03.04

# Define run parameters and input files
R1_FILE=$(readlink -f ERR346600_1.fastq)
R2_FILE=$(readlink -f ERR346600_2.fastq)
R1_PRIMERS=$(readlink -f Greiff2014_CPrimers.fasta)
R2_PRIMERS=$(readlink -f Greiff2014_VPrimers.fasta)
OUTDIR="output"
OUTNAME="M1"
NPROC=4
PIPELINE_LOG="Pipeline.log"
ZIP_FILES=true

# Make output directory and empty log files
mkdir -p $OUTDIR; cd $OUTDIR
echo '' > $PIPELINE_LOG

# Start
echo "OUTPUT DIRECTORY: ${OUTDIR}"
echo -e "START"
STEP=0

# Assemble paired ends via mate-pair alignment
printf "  %2d: %-*s $(date +'%H:%M %D')\n" $((++STEP)) 24 "AssemblePairs align"
AssemblePairs.py align -1 $R2_FILE -2 $R1_FILE --coord sra --rc tail \
	--outname $OUTNAME --outdir . --log AP.log --nproc $NPROC >> $PIPELINE_LOG

# Remove low quality reads
printf "  %2d: %-*s $(date +'%H:%M %D')\n" $((++STEP)) 24 "FilterSeq quality"
FilterSeq.py quality -s "${OUTNAME}_assemble-pass.fastq" -q 20 \
	--outname $OUTNAME --log FS.log --nproc $NPROC >> $PIPELINE_LOG

# Identify forward and reverse primers
printf "  %2d: %-*s $(date +'%H:%M %D')\n" $((++STEP)) 24 "MaskPrimers score"
MaskPrimers.py score -s "${OUTNAME}_quality-pass.fastq" -p $R2_PRIMERS \
	--mode mask --start 4 --outname "${OUTNAME}-FWD" --log MPV.log \
	--nproc $NPROC >> $PIPELINE_LOG
MaskPrimers.py score -s "${OUTNAME}-FWD_primers-pass.fastq" -p $R1_PRIMERS \
	--mode cut --start 4 --revpr --outname "${OUTNAME}-REV" --log MPC.log \
	--nproc $NPROC >> $PIPELINE_LOG

# Expand primer field
printf "  %2d: %-*s $(date +'%H:%M %D')\n" $((++STEP)) 24 "ParseHeaders expand"
ParseHeaders.py expand -s "${OUTNAME}-REV_primers-pass.fastq" \
    -f PRIMER >> $PIPELINE_LOG

# Rename primer fields
printf "  %2d: %-*s $(date +'%H:%M %D')\n" $((++STEP)) 24 "ParseHeaders rename"
ParseHeaders.py rename -s "${OUTNAME}-REV_primers-pass_reheader.fastq" \
    -f PRIMER1 PRIMER2 -k VPRIMER CPRIMER --outname $OUTNAME >> $PIPELINE_LOG
	
# Remove duplicate sequences
printf "  %2d: %-*s $(date +'%H:%M %D')\n" $((++STEP)) 24 "CollapseSeq"
CollapseSeq.py -s "${OUTNAME}_reheader.fastq" -n 20 \
	--uf CPRIMER --cf VPRIMER --act set --inner \
	--outname $OUTNAME >> $PIPELINE_LOG

# Subset to sequences observed at least twice
printf "  %2d: %-*s $(date +'%H:%M %D')\n" $((++STEP)) 24 "SplitSeq group"
SplitSeq.py group -s "${OUTNAME}_collapse-unique.fastq" -f DUPCOUNT --num 2 \
	--outname $OUTNAME >> $PIPELINE_LOG

# Create annotation table of final unique sequences
printf "  %2d: %-*s $(date +'%H:%M %D')\n" $((++STEP)) 24 "ParseHeaders table"
ParseHeaders.py table -s "${OUTNAME}_atleast-2.fastq" \
    -f ID DUPCOUNT CPRIMER VPRIMER >> $PIPELINE_LOG

# Process log files
printf "  %2d: %-*s $(date +'%H:%M %D')\n" $((++STEP)) 24 "ParseLog"
ParseLog.py -l AP.log -f ID LENGTH OVERLAP ERROR PVALUE > /dev/null &
ParseLog.py -l FS.log -f ID QUALITY > /dev/null &
ParseLog.py -l MP[VC].log -f ID PRIMER ERROR > /dev/null &
wait

# Zip intermediate and log files
if $ZIP_FILES; then
    LOG_FILES_ZIP=$(ls AP.log FS.log MP[VC].log)
    tar -zcf LogFiles.tar.gz $LOG_FILES_ZIP
    rm $LOG_FILES_ZIP

    TEMP_FILES_ZIP=$(ls *.fastq | grep -vP "collapse-unique.fastq|atleast-2.fastq")
    tar -zcf TempFiles.tar.gz $TEMP_FILES_ZIP
    rm $TEMP_FILES_ZIP
fi

# End
printf "DONE\n\n"
cd ../


RScripts/DiversityAnalysis.R
library(alakazam)

# Load Change-O file
df <- readChangeoDb('/home/ian/NextGenSeq/Greiff2014/output/HighVQuestOutput/Greiff2014_db-pass_parse-select_clone-pass_germ-pass.tab')

# Partitions the data based on the SAMPLE column
clones <- countClones(df, groups="CPRIMER")
head(clones, 5)

# Partitions the data based on both the SAMPLE and ISOTYPE columns
# Weights the clone sizes by the DUPCOUNT column
clones <- countClones(df, groups="CPRIMER", copy="DUPCOUNT")
head(clones, 5)

# Partitions the data on the SAMPLE column
# Calculates a 95% confidence interval via 200 bootstrap realizations
clones <- estimateAbundance(df, group="CPRIMER", ci=0.95, nboot=200)

# Plots a rank abundance curve of the relative clonal abundances
p1 <- plotAbundance(clones)

# Compare diversity curve across values in the "SAMPLE" column
# q ranges from 0 (min_q=0) to 32 (max_q=32) in 0.05 incriments (step_q=0.05)
# A 95% confidence interval will be calculated (ci=0.95)
# 2000 resampling realizations are performed (nboot=200)
sample_div <- rarefyDiversity(df, "CPRIMER", min_q=0, max_q=32, step_q=0.05, 
                              ci=0.95, nboot=200)

# Plot a log-log (log_q=TRUE, log_d=TRUE) plot of sample diversity
# Indicate number of sequences resampled from each group in the title
sample_main <- paste0("Sample diversity ")
p2 <- plotDiversityCurve(sample_div, main_title=sample_main, 
                         legend_title="Sample", log_q=TRUE, log_d=TRUE)

RScripts/IgLineageReconstruction.R
library(alakazam)
library(igraph)
library(dplyr)

# Load Change-O file
#file <- system.file("extdata", "ExampleDb.gz", package="alakazam")
df <- readChangeoDb('/home/ian/NextGenSeq/Greiff2014_2/output/HighVQuestOutput/Greiff2014_db-pass_parse-select_clone-pass_germ-pass.tab')
#df <- readChangeoDb('/home/ian/NextGenSeq/Greiff2014_2/output/ChangeO/Greiff2014_db-pass_parse-select_clone-pass_germ-pass.tab')

# Select clone
sub_df <- subset(df, CLONE == 380)

# This example data set does not have ragged ends
# Preprocess clone without ragged end masking (default)
clone <- makeChangeoClone(sub_df, 
                          num_fields="DUPCOUNT")

# Show combined annotations
clone@data[, c("DUPCOUNT")]

# Run PHYLIP and parse output
dnapars_exec <- "/home/ian/BioinformaticsTools/phylip-3.696/exe/dnapars"
graph <- buildPhylipLineage(clone, dnapars_exec, rm_temp=TRUE)

# The graph has shared annotations for the clone
data.frame(CLONE=graph$clone,
           JUNCTION_LENGTH=graph$junc_len,
           V_GENE=graph$v_gene,
           J_GENE=graph$j_gene)

# The vertices have sequence specific annotations
data.frame(SEQUENCE_ID=V(graph)$name, 
           DUPCOUNT=V(graph)$DUPCOUNT)

# Plot graph with defaults
plot(graph)

# Modify graph and plot attributes
V(graph)$color <- "lightgrey"
V(graph)$color[V(graph)$name == "Germline"] <- "black"
V(graph)$color[grepl("Inferred", V(graph)$name)] <- "white"
E(graph)$label <- ""

# Remove large default margins
#par(mar=c(0, 0, 0, 0) + 0.1)

# Define a tree layout with the Germline at the top
ly <- layout_as_tree(graph, root="Germline", circular=F, flip.y=T)

# Plot graph
plot(graph, layout=ly, edge.arrow.mode=0, vertex.frame.color="black",
     vertex.label.color="black", vertex.size=30)

# Add legend
legend("topleft", c("Germline", "Inferred", "Sample"), 
       fill=c("black", "white", "grey80"), cex=0.75)

Pipeline_Greiff2014_fasta.sh
#!/usr/bin/env bash
# Example pRESTO pipeline for Stranded MiSeq 2x250 data.
# Data from Greiff et al, 2014, BMC Immunol.
#
# Author:  Jason Anthony Vander Heiden
# Date:    2016.03.04

# Define run parameters and input files
R1_FILE=$(readlink -f ERR346600_1.fastq)
R2_FILE=$(readlink -f ERR346600_2.fastq)
R1_PRIMERS=$(readlink -f Greiff2014_CPrimers.fasta)
R2_PRIMERS=$(readlink -f Greiff2014_VPrimers.fasta)
OUTDIR="output"
OUTNAME="M1"
NPROC=4
PIPELINE_LOG="Pipeline.log"
ZIP_FILES=true

# Make output directory and empty log files
mkdir -p $OUTDIR; cd $OUTDIR
echo '' > $PIPELINE_LOG

# Start
echo "OUTPUT DIRECTORY: ${OUTDIR}"
echo -e "START"
STEP=0

# Assemble paired ends via mate-pair alignment
printf "  %2d: %-*s $(date +'%H:%M %D')\n" $((++STEP)) 24 "AssemblePairs align"
AssemblePairs.py align -1 $R2_FILE -2 $R1_FILE --coord sra --rc tail \
	--outname $OUTNAME --outdir . --log AP.log --nproc $NPROC >> $PIPELINE_LOG

# Remove low quality reads
printf "  %2d: %-*s $(date +'%H:%M %D')\n" $((++STEP)) 24 "FilterSeq quality"
FilterSeq.py quality -s "${OUTNAME}_assemble-pass.fastq" -q 20 \
	--outname $OUTNAME --log FS.log --nproc $NPROC >> $PIPELINE_LOG

# Identify forward and reverse primers
printf "  %2d: %-*s $(date +'%H:%M %D')\n" $((++STEP)) 24 "MaskPrimers score"
MaskPrimers.py score -s "${OUTNAME}_quality-pass.fastq" -p $R2_PRIMERS \
	--mode mask --start 4 --outname "${OUTNAME}-FWD" --log MPV.log \
	--nproc $NPROC >> $PIPELINE_LOG
MaskPrimers.py score -s "${OUTNAME}-FWD_primers-pass.fastq" -p $R1_PRIMERS \
	--mode cut --start 4 --revpr --outname "${OUTNAME}-REV" --log MPC.log \
	--nproc $NPROC >> $PIPELINE_LOG

# Expand primer field
printf "  %2d: %-*s $(date +'%H:%M %D')\n" $((++STEP)) 24 "ParseHeaders expand"
ParseHeaders.py expand -s "${OUTNAME}-REV_primers-pass.fastq" --fasta \
    -f PRIMER >> $PIPELINE_LOG

# Rename primer fields
printf "  %2d: %-*s $(date +'%H:%M %D')\n" $((++STEP)) 24 "ParseHeaders rename"
ParseHeaders.py rename -s "${OUTNAME}-REV_primers-pass_reheader.fasta" --fasta \
    -f PRIMER1 PRIMER2 -k VPRIMER CPRIMER --outname $OUTNAME >> $PIPELINE_LOG

# Remove duplicate sequences
printf "  %2d: %-*s $(date +'%H:%M %D')\n" $((++STEP)) 24 "CollapseSeq"
CollapseSeq.py -s "${OUTNAME}_reheader.fasta" -n 15 \
	--uf CPRIMER --cf VPRIMER --act set --inner \
	--outname $OUTNAME >> $PIPELINE_LOG

# Subset to sequences observed at least twice
printf "  %2d: %-*s $(date +'%H:%M %D')\n" $((++STEP)) 24 "SplitSeq group"
SplitSeq.py group -s "${OUTNAME}_collapse-unique.fasta" -f DUPCOUNT --num 2 \
	--outname $OUTNAME >> $PIPELINE_LOG

# Create annotation table of final unique sequences
printf "  %2d: %-*s $(date +'%H:%M %D')\n" $((++STEP)) 24 "ParseHeaders table"
ParseHeaders.py table -s "${OUTNAME}_atleast-2.fasta" \
    -f ID DUPCOUNT CPRIMER VPRIMER >> $PIPELINE_LOG

# Process log files
printf "  %2d: %-*s $(date +'%H:%M %D')\n" $((++STEP)) 24 "ParseLog"
ParseLog.py -l AP.log -f ID LENGTH OVERLAP ERROR PVALUE > /dev/null &
ParseLog.py -l FS.log -f ID QUALITY > /dev/null &
ParseLog.py -l MP[VC].log -f ID PRIMER ERROR > /dev/null &
wait

# Zip intermediate and log files
if $ZIP_FILES; then
    LOG_FILES_ZIP=$(ls AP.log FS.log MP[VC].log)
    tar -zcf LogFiles.tar.gz $LOG_FILES_ZIP
    rm $LOG_FILES_ZIP

    TEMP_FILES_ZIP=$(ls *.fasta | grep -vP "collapse-unique.fasta|atleast-2.fasta")
    tar -zcf TempFiles.tar.gz $TEMP_FILES_ZIP
    rm $TEMP_FILES_ZIP
fi

# End
printf "DONE\n\n"
cd ../
